operator/standalone/delete.go

```diff
@@ -0,0 +1,77 @@
+	"github.com/vshn/appcat-service-postgresql/apis/postgresql/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	pipeline "github.com/ccremer/go-command-pipeline"
+)
+
+const crossplaneManagedFinalizer = "finalizer.managedresource.crossplane.io"
+
+// DeleteStandalonePipeline is a pipeline that deletes an instance from the target deployment namespace.
+type DeleteStandalonePipeline struct {
+	client   client.Client
+	instance *v1alpha1.PostgresqlStandalone
+}
+
+// NewDeleteStandalonePipeline creates a new delete pipeline with the required dependencies.
+func NewDeleteStandalonePipeline(client client.Client, instance *v1alpha1.PostgresqlStandalone) *DeleteStandalonePipeline {
+	return &DeleteStandalonePipeline{
+		instance: instance,
+		client:   client,
+	}
+}
+
+// RunPipeline executes the pipeline with configured business logic steps.
+func (d *DeleteStandalonePipeline) RunPipeline(ctx context.Context) error {
+	return pipeline.NewPipeline().
+		WithSteps(
+			pipeline.NewStepFromFunc("delete helm release", d.deleteHelmRelease),
+			pipeline.NewStepFromFunc("delete namespace", d.deleteNamespace),
+			pipeline.NewStepFromFunc("delete connection secret", d.deleteConnectionSecret),
+		).
+		RunWithContext(ctx).Err()
+}
+
+// deleteHelmRelease removes the Helm Release from the cluster
+// The crossplaneManagedFinalizer is removed before the Release is deleted
+func (d *DeleteStandalonePipeline) deleteHelmRelease(ctx context.Context) error {
+	helmRelease := &helmv1beta1.Release{}
+	err := d.client.Get(ctx, client.ObjectKey{Name: d.instance.Status.HelmChart.DeploymentNamespace}, helmRelease)
+	if err != nil {
+		return err
+	}
+	removed := controllerutil.RemoveFinalizer(helmRelease, crossplaneManagedFinalizer)
+	if !removed {
+		return fmt.Errorf("cannot remove finalizer %s from helm release %s", crossplaneManagedFinalizer, helmRelease.Name)
+	}
+	// TODO add DeletionPropagation ???
+	return d.client.Delete(ctx, helmRelease)
+}
+
+// deleteNamespace removes the namespace of the PostgreSQL instance
+func (d *DeleteStandalonePipeline) deleteNamespace(ctx context.Context) error {
+	deploymentNamespace := &corev1.Namespace{}
+	err := d.client.Get(ctx, client.ObjectKey{Name: d.instance.Status.HelmChart.DeploymentNamespace}, deploymentNamespace)
+	if err != nil {
+		return err
+	}
+	return d.client.Delete(ctx, deploymentNamespace)
+}
+
+// deleteConnectionSecret removes the connection secret of the PostgreSQL instance
+func (d *DeleteStandalonePipeline) deleteConnectionSecret(ctx context.Context) error {
+	connectionSecret := &corev1.Secret{}
+	err := d.client.Get(ctx, types.NamespacedName{Name: d.instance.Spec.WriteConnectionSecretToRef.Name, Namespace: d.instance.Namespace}, connectionSecret)
+	if err != nil {
+		return err
+	}
+	return d.client.Delete(ctx, connectionSecret)
```

Comment:
same here.

Reply:
Let me check if setting just the name will do the job (and namespace)

Reply:
I don't see a reason why it shouldn't.
Actually, theoretically speaking it might not even be needed, since I put an owner reference to the instance object in the secret. That way, kubernetes should already remove it automatically. But maybe it makes sense to be sure?

Reply:
it's better to be sure in this case, I updated the code and removed the `get` method from all 3 resources.